src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/AssignVacancyNumberCommandHandler.cs

```diff
@@ -0,0 +1,47 @@
+﻿using Esfa.Recruit.Vacancies.Client.Application.Commands;
+using Esfa.Recruit.Vacancies.Client.Application.Services;
+using Esfa.Recruit.Vacancies.Client.Domain.Messaging;
+using Esfa.Recruit.Vacancies.Client.Domain.Repositories;
+using MediatR;
+using Microsoft.Extensions.Logging;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Esfa.Recruit.Vacancies.Client.Application.CommandHandlers
+{
+    public class AssignVacancyNumberCommandHandler: IRequestHandler<AssignVacancyNumberCommand>
+    {
+        private readonly IVacancyRepository _repository;
+        private readonly IMessaging _messaging;
+        private readonly ILogger<AssignVacancyNumberCommandHandler> _logger;
+        private readonly IGenerateVacancyNumbers _generator;
+
+        public AssignVacancyNumberCommandHandler(
+            IVacancyRepository repository, 
+            IMessaging messaging, 
+            ILogger<AssignVacancyNumberCommandHandler> logger, 
+            IGenerateVacancyNumbers generator)
+        {
+            _repository = repository;
+            _messaging = messaging;
+            _logger = logger;
+            _generator = generator;
+        }
+
+        public async Task Handle(AssignVacancyNumberCommand message, CancellationToken cancellationToken)
+        {
+            var vacancy = await _repository.GetVacancyAsync(message.VacancyId);
+            
+            if (vacancy.VacancyReference.HasValue)
+            {
+                _logger.LogWarning("Vacancy: {vacancyId} already has a vacancy number: {vacancyNumber}. Will not be changed.", vacancy.Id, vacancy.VacancyReference);
+            }
+
+            vacancy.VacancyReference = await _generator.GenerateAsync();
+
+            await _repository.UpdateAsync(vacancy);
+
+            _logger.LogInformation("Updated Vacancy: {vacancyId} with vacancy number: {vacancyNumber}", vacancy.Id, vacancy.VacancyReference);
```

Comment:
missing $ maybe?

Reply:
Nope. Logger uses string format for structured logging. i.e. vacancyId and vacancyNumber will appear in Kibana as properties of the log message as well as in the message string itself.